Add flexible bilinear upsampling aspect ratio redux

andrewgiessel opened the pull request:
This PR addresses issue #1257, which aims to add non-coupled scaling factors for bilinear 2d upsampling. I've added two new tests, and currently everything passes. This is a relatively simple change which only touches python code.

Furthermore, this is is a re-do of PR https://github.com/pytorch/pytorch/pull/1279, wherein I messed up a rebase.  There are some in-line comments there that might be worth skimming through, as well, but I tried to address most of the concerns raised there.  In particular, I maintained the base class of all the upsampling methods.

Tests pass, on GPU and CPU.

tagging @apaszke @fmassa @soumith.  Thanks in advance everyone.

ps:  I'll probably need a pointer on how to best rebase to master, eventually ;)

Commit:
bilinear upsampling module

Commit:
bilinear upsampling functions

Commit:
bilinear upsampling tests

apaszke commented on torch/nn/_functions/thnn/upsampling.py:
```diff
@@ -12,16 +13,20 @@ def __init__(self, size=None, scale_factor=None):
-        self.size = size
+        if scale_factor is not None and not isinstance(scale_factor, (Integral, tuple)):
+            raise ValueError('scale_factor must be of integer type or tuple of integer types')
+        self.size = _pair(size)
```
It should be `_pair(size) if size is not None else None`

apaszke commented on torch/nn/_functions/thnn/upsampling.py:
```diff
@@ -64,13 +69,29 @@ def backward(self, grad_output):
+            # we have to be a tuple at this point
+            try:
+                assert len(self.scale_factor) == 2
+                for i in self.scale_factor:
```
```python
assert all(isinstance(s, Integral) and s >= 1 for s in self.scale_factor)
```

apaszke commented on torch/nn/_functions/thnn/upsampling.py:
```diff
@@ -64,13 +69,29 @@ def backward(self, grad_output):
             self.output_size = (
-                input.size(2) * self.scale_factor,
-                input.size(3) * self.scale_factor,
+                input.size(2) * self.scale_factor[0],
```
You need to write a `__setstate__` function for modules that were serialized - `self.scale_factor` won't be a tuple for them, so you need to patch it:
```python
def __setstate__(self, state):
    self.__dict__.update(state)
    self.scale_factor = _tuple(self.scale_factor)
```

apaszke commented on torch/nn/modules/upsampling.py:
```diff
@@ -110,5 +115,21 @@ class UpsamplingBilinear2d(_UpsamplingBase):
     """
 
+    def __init__(self, size=None, scale_factor=None):
+        super(UpsamplingBilinear2d, self).__init__(size, scale_factor)
```
As said in an earlier PR - can you move these checks to a helper function (prefixed with an underscore) in `functional.py` and call it from here?

apaszke commented on torch/nn/modules/upsampling.py:
```diff
@@ -11,8 +11,8 @@ def __init__(self, size=None, scale_factor=None):
-            raise ValueError('scale_factor must be of integer type')
+        if scale_factor is not None and not isinstance(scale_factor, (Integral, tuple)):
+            raise ValueError('scale_factor must be of integer type or tuple of integer types')
         self.size = _pair(size)
```
I think it also needs that if `self.size is not None`

apaszke commented on torch/nn/_functions/thnn/upsampling.py:
```diff
@@ -64,13 +69,29 @@ def backward(self, grad_output):
+
+        if self.scale_factor is not None:
+            self.scale_factor = _pair(self.scale_factor)
+            # we have to be a tuple at this point
```
That comment seems is unnecessary

Commit:
refactor out bilinear_scale_factor check

Moved to functional.py

Commit:
checks for None in _pair() calls

Commit:
adding __setstate__()

andrewgiessel commented:
Thanks for the comments @apaszke! I think I made all the changes you asked for.  Please, let me know if I didn't understand something (in particular I just c/p your `__setstate__()` function).

Commit:
Adding Integral import

Commit:
actually return scale_factor after check

Commit:
sneak linting

Commit:
Move casting to tuples out of parent function

This allows for the base class to be used for upsampling routines other
than 2d.  I also renamed _check_bilinear_2d_scale_factor().

andrewgiessel commented:
Please see the comment I made [here](https://github.com/pytorch/pytorch/pull/1348#issuecomment-297800836) on #1348 explaining the most recent commit.

soumith commented:
thanks Andrew!
